### PR TITLE
Add CountAtLeast set

### PR DESCRIPTION
**Pull request by odow**
Part of #1805

Stacked on #1827 to avoid conflicts.

### CODE REVIEW

File: test/sets.jl [hunk 1]
```diff
@@ -156,6 +156,7 @@ function test_sets_DimensionMismatch()
     @test_throws DimensionMismatch MOI.Complements(-3)
     @test_throws DimensionMismatch MOI.Complements(3)
     @test_throws DimensionMismatch MOI.Among(0, Set([1, 2]))
+    @test_throws DimensionMismatch MOI.CountAtLeast(0, 2, Set([1, 2]))
```
Comment (odow) [inline test/sets.jl:159]:
```suggestion
    @test_throws DimensionMismatch MOI.CountAtLeast(-1, 2, Set([1, 2]))
```

Comment (odow) [inline test/sets.jl:159]:
```suggestion
    @test_throws DimensionMismatch MOI.CountAtLeast(-1, 2, Set([1, 2]))
```

File: src/Test/test_cpsat.jl [hunk 2]
```diff
@@ -122,3 +122,45 @@ function setup_test(
...
+        model,
+        MOI.VectorOfVariables,
+        MOI.CountAtLeast,
+    )
+    @requires MOI.supports_add_constrained_variable(model, MOI.Integer)
+    @requires _supports(config, MOI.optimize!)
+    y = [MOI.add_constrained_variable(model, MOI.Integer()) for _ in 1:3]
+    x = first.(y)
+    set = Set([3, 4])
+    MOI.add_constraint(model, MOI.VectorOfVariables(x), MOI.Among(3, 2, set))
```
Comment (odow) [inline src/Test/test_cpsat.jl:145]:
```suggestion
    MOI.add_constraint(
        model,
        MOI.VectorOfVariables(x),
        MOI.CountAtLeast(3, 2, set),
)
```

File: src/Test/test_cpsat.jl [hunk 3]
```diff
@@ -122,3 +122,49 @@ function setup_test(
...
+    @requires MOI.supports_add_constrained_variable(model, MOI.Integer)
+    @requires _supports(config, MOI.optimize!)
+    y = [MOI.add_constrained_variable(model, MOI.Integer()) for _ in 1:3]
+    x = first.(y)
+    set = Set([3, 4])
+    MOI.add_constraint(
+        model,
+        MOI.VectorOfVariables(x),
+        MOI.CountAtLeast(3, 2, set),
+)
```
Comment (odow) [inline src/Test/test_cpsat.jl:149]:
```suggestion
    )
```

**Comment by odow** [comment]
Okay, this definition doesn't quite match the MiniZinc one, because it's `array [$X] of var set of int: x`, not `array [$X] of var int: x`. Note the additional "set of":
```Julia
include "globals.mzn";
var 1 .. 3: x1;
var 1 .. 3: x2;
var 1 .. 3: x3;
constraint at_least(1, [{x1,x2}, {x2,x3}], {2});
solve satisfy;
```
We just need to add a new field with the partitioning of the elements in `x`.

File: test/sets.jl [hunk 4]
```diff
@@ -156,6 +156,7 @@ function test_sets_DimensionMismatch()
     @test_throws DimensionMismatch MOI.Complements(-3)
     @test_throws DimensionMismatch MOI.Complements(3)
     @test_throws DimensionMismatch MOI.Among(0, Set([1, 2]))
+    @test_throws DimensionMismatch MOI.CountAtLeast(-1, 2, Set([1, 2]))
```
Comment (odow) [inline test/sets.jl:159]:
```suggestion
    @test_throws DimensionMismatch MOI.CountAtLeast([-1, 2], Set([1, 2]))
```

File: src/sets.jl [hunk 5]
```diff
@@ -1216,6 +1216,87 @@ end
 
 Base.:(==)(x::Among, y::Among) = x.dimension == y.dimension && x.set == y.set
 
+"""
+    CountAtLeast(n::Int, partitions::Vector{Int}, set::Set{Int})
+
+The set ``\\{x \\in \\mathbb{R}^{d_1 + d_2 + \\ldots d_N}\\}``, where `x` is
+partitioned into `N` subsets (``\\{x_1,  \\ldots, x_{d_1}\\}``,
+``\\{x_{1 + d_1},  \\ldots, x_{d_1 + d_2}\\}`` and so on), and at least ``n``
```
Comment (blegat) [inline src/sets.jl:1224]:
```suggestion
``\\{x_{d_1 + 1},  \\ldots, x_{d_1 + d_2}\\}`` and so on), and at least ``n``
```

File: src/sets.jl [hunk 6]
```diff
@@ -1216,6 +1216,87 @@ end
 
 Base.:(==)(x::Among, y::Among) = x.dimension == y.dimension && x.set == y.set
 
+"""
+    CountAtLeast(n::Int, partitions::Vector{Int}, set::Set{Int})
```
Comment (blegat) [inline src/sets.jl:1220]:
```suggestion
    CountAtLeast(n::Int, d::Vector{Int}, set::Set{Int})
```

File: src/sets.jl [hunk 7]
```diff
@@ -1216,6 +1216,87 @@ end
...
+
+The set ``\\{x \\in \\mathbb{R}^{d_1 + d_2 + \\ldots d_N}\\}``, where `x` is
+partitioned into `N` subsets (``\\{x_1,  \\ldots, x_{d_1}\\}``,
+``\\{x_{1 + d_1},  \\ldots, x_{d_1 + d_2}\\}`` and so on), and at least ``n``
+elements of each subset take one of the values in `set`.
+
+This constraint is sometimes called `at_least`.
+
+## Example
+
```
Comment (blegat) [inline src/sets.jl:1230]: `3` appears at least once in `{x, y}` and `{y, z}`

File: src/sets.jl [hunk 8]
```diff
@@ -1216,6 +1216,87 @@ end
 
 Base.:(==)(x::Among, y::Among) = x.dimension == y.dimension && x.set == y.set
 
+"""
+    CountAtLeast(n::Int, partitions::Vector{Int}, set::Set{Int})
+
+The set ``\\{x \\in \\mathbb{R}^{d_1 + d_2 + \\ldots d_N}\\}``, where `x` is
+partitioned into `N` subsets (``\\{x_1,  \\ldots, x_{d_1}\\}``,
```
Comment (blegat) [inline src/sets.jl:1223]: What happens if a variable appears twice in the same subset ?

**Pull request merged**